ui/timeline: allow subscribing to UTDs and late-decrypt events

bnjbvr opened the pull request:
This adds a new mechanism in the UI crate (since re-attempts to decrypt happen in the timeline, as of today — later that'll happen in the event cache) to notify whenever we run into a UTD (an event couldn't be decrypted) or a late-decryption event (after some time, a UTD could be decrypted).

This new hook will deduplicate pings for the same event (identifying events on their event id), and also implements an optional grace period. If an event was a UTD:

- if it's still a UTD after the grace period, then it's reported with a `None` `time_to_decrypt`,
- if it's not a UTD anymore (i.e. it's been decrypted in the meanwhile), then it's reported with a `time_to_decrypt` set to the time it took to decrypt the event (approximate, since it starts counting from the time the timeline receives it, not the time the SDK fails to decrypt it at first).

It's configurable as an optional hook on timeline builders. For the FFI, it's configurable at the sync service's level with a "delegate", and then the sync service will forward the hook to the timelines it creates, and the hook will forward the UTD info to the delegate.

Part of https://github.com/element-hq/element-meta/issues/2300.

stefanceriu commented on crates/matrix-sdk-ui/src/unable_to_decrypt_hook.rs:
```diff
@@ -0,0 +1,370 @@
+/// Information about an event we were unable to decrypt (UTD).
+#[derive(Clone, Debug, Hash, PartialEq, Eq)]
+#[cfg_attr(feature = "uniffi", derive(uniffi::Object))]
+pub struct UnableToDecryptInfo {
```
Hmm.. these fields are not exposed to Swift.

bnjbvr replied:
and the facepalm of the week goes to yours truly!

stefanceriu commented on crates/matrix-sdk-ui/src/unable_to_decrypt_hook.rs:
```diff
@@ -0,0 +1,370 @@
+#[cfg_attr(feature = "uniffi", derive(uniffi::Object))]
+pub struct UnableToDecryptInfo {
+    /// The identifier of the event that couldn't get decrypted.
+    pub event_id: OwnedEventId,
```
On a related note, I dont think this information is enough to do what's requested in the ticket and what events are available in analytics: https://github.com/element-hq/element-meta/issues/2300

stefanceriu replied:
Okay, we agreed that for now we can hardcode all of them to `OlmKeysNotSentError`

bmarty commented:
Testing on the Android side, after building commit 7dc0bbdf32d70085fb80c8fe85a60a8155872cb1
- I confirm that `on_utd()` is called 60s after an UTD Event has been received by the session 👍 
- I do not see any `on_late_decrypt()` method in the API
- No fields are exposed in `UnableToDecryptInfo`.

Here is the draft PR on EXA: https://github.com/element-hq/element-x-android/pull/2544

bnjbvr commented on crates/matrix-sdk-ui/src/unable_to_decrypt_hook.rs:
```diff
@@ -0,0 +1,370 @@
+        assert_eq!(wrapper.pending_delayed.lock().unwrap().len(), 1);
+
+        // If I wait for 1 second, then it's still not been notified yet.
+        sleep(Duration::from_secs(1)).await;
```
That's a great idea! As discussed, will attempt to do this in a follow-up PR :innocent:

andybalaam replied:
Awesome, thank you!

bnjbvr commented on bindings/matrix-sdk-ffi/src/sync_service.rs:
```diff
@@ -101,17 +110,64 @@ impl SyncServiceBuilder {
+        // UTDs that have been decrypted in the `IGNORE_UTD_PERIOD` are just ignored and
+        // not considered UTDs.
+        if let Some(duration) = &info.time_to_decrypt {
+            if *duration < IGNORE_UTD_PERIOD {
```
The ignore part really felt like an EX-specific ask (which the FFI layer fits well), when the grace period felt more useful in general. Happy to move it to the UI layer if you think strongly about it, I could add some tests for it too, but it seemed simple enough to not mandate any?

andybalaam replied:
I'm OK with its locations - I was just asking. I would love tests for it, because I love tests for everything.

bnjbvr commented on crates/matrix-sdk-ui/src/unable_to_decrypt_hook.rs:
```diff
@@ -0,0 +1,370 @@
+
+        // Spawn a task that will wait for the given delay, and maybe call the parent
+        // hook then.
+        let handle = spawn(async move {
```
Indeed; as opposed to the `known_utds` data structure, they should also degrow over time, so it's really a function of the number of events we receive. Will add a note.

bnjbvr commented:
Addressed review comments. In particular there's now a new FFI `UnableToDecryptInfo` data structure correctly exposing the public fields. @andybalaam PTAL!

andybalaam commented on crates/matrix-sdk-ui/src/unable_to_decrypt_hook.rs:
```diff
@@ -0,0 +1,370 @@
+        }
+
+        // And when I call the `on_late_decrypt` method before the event had been marked
+        // as utd,
```
But is the comment right? I thought we had already marked this as UTD. Maybe I'm reading it wrong.

bnjbvr replied:
The comment was indeed wrong, I fixed it — sorry, forgot to update here!

andybalaam reviewed (approved):
Looks good. Everything else is optional from my POV. Thanks!
